### ISSUE TITLE
Various Typescript Definitions Changes Needed

### ISSUE THREAD

**Issue by bradennapier** [sindresorhus/got#1331]
#### What problem are you trying to solve?

Thanks for the library!  

The typescript types are pretty problematic in general - especially when trying to use them outside of directly calling the functions (such as helper functions or libraries which want to use got).  In some cases there are places that the types are completely invalid as well.

These are what I have run into thus far, I will update as I find more but in general it seems like some changes would help make the library work in more cases without casting or other hackery.

> I believe a leading cause for the issues not being noticed during development is not using `strict` / `strictNullChecks` in the `tsconfig.json` compileOptions.

#### 1. `searchParams` does not allow `undefined` 

This makes it impossible to do a lot of situations where you need to build up the params object without casting to any 

![image](https://user-images.githubusercontent.com/15365418/85915631-3c152a00-b817-11ea-9e20-997db9725fa6.png)

```typescript
// the goal is to add `starting_after` to the params object after the initial request in a 
// generator.  
class SomeClass {
  async *getContactsPaginated(perPage: number, maxPage = Infinity) {
    let page: IntercomPagesCursor;
    const searchParams: { per_page: number; starting_after?: string } = {
      per_page: perPage,
    };
    while (true) {
      const result = await this.client.get<
        IntercomCursorPaginatedList<Contact>
      >('contacts', { searchParams });

      page = result.pages;
      
      yield result.data;

      if (!page || !page.next || page.page >= maxPage) {
        break;
      }

      searchParams.starting_after = page.next.starting_after;
    }
  }
}

async function run() {
  for await (const contacts of client.getContactsPaginated(25, 5)) {
    // handle contacts
  }
}
```

> Basically in most areas that index types are provided it should also have `| undefined` so that types can be defined in this way.  `searchParams?: string | { [key: string]: string | number | boolean | null | undefined; } | URLSearchParams;`

#### 2. `Options` is impossible to derive in a way that doesn't cause errors

There is no way to create a function with a type that got will allow for options. For example, when a helper is being used
in-between a rest api and the user.

![image](https://user-images.githubusercontent.com/15365418/85915643-4800ec00-b817-11ea-8523-5e882330d122.png)

```typescript
 async rawRequest<R>(
    method: HTTPAlias,
    path: string,
    // THIS IS NOT ACCEPTED BY got.get(path, options) with `as never` or `as any` 
    options?: GotOptions,
  ): Promise<GotResponse<R>> {
    try {
      const result = await this.client[method]<R>(path, options);
      return result;
    } catch (error) {
      handleHTTPError(error);
      throw error;
    }
  }
```

> I have tried every possible method of deriving such as `Parameters<typeof Got['get']>[1]` all of which fail due to the way the type was built

> I did notice that in the `types.d.ts` file it imports `Options as RequestOptions` then the exported `Options` extends that object further.  If I export the `RequestOptions` and use that instead of `Options` then it seems to work, but that is not exported as far as I can tell otherwise and I am not sure if there are other implications of that. Or have to do hacky import like `import { Options as GotRequestOptions } from 'got/dist/source/core';`

### Checklist

- [x] I have read the documentation and made sure this feature doesn't already exist.

**Comment by sindresorhus** [comment]
I agree, the TS types need to be improved. Help is always welcome with that.

---

> I believe a leading cause for the issues not being noticed during development is not using strict / strictNullChecks in the tsconfig.json compileOptions.

We are using all the strict options.

> searchParams does not allow undefined

Yeah, I guess we could allow `undefined` too and just ignore the key then.

However, note that the browser type for `URLSearchParams` doesn't allow either null or undefined: https://github.com/microsoft/TypeScript/blob/9c71eaf59040ae75343da8cdff01344020f5bba2/src/lib/dom.generated.d.ts#L14606

> Options is impossible to derive in a way that doesn't cause errors

This is a long-standing issue I really like to see resolved. It's partly our fault and partly TS for being buggy.

**Comment by bradennapier** [comment]
> However, note that the browser type for `URLSearchParams` doesn't allow either null or undefined: https://github.com/microsoft/TypeScript/blob/9c71eaf59040ae75343da8cdff01344020f5bba2/src/lib/dom.generated.d.ts#L14606

This is because of how you create it.  You don't provide an object for it you build it `new URLSearchParams('?one=two')` and/or then use get/set directly so it wouldn't need that.  Adding the `undefined` is probably still ideal but in my case I could just use `URLSearchParams` as well.  The diff is it isn't as type safe since in my case I am strictly only allowed to use those two parameters.

**Comment by bradennapier** [comment]
Actually that doesn't work out of the box:

![image](https://user-images.githubusercontent.com/15365418/85916832-50aaef80-b822-11ea-876f-ce962853c760.png)


```
 Type 'URLSearchParams' is not assignable to type 'import("url").URLSearchParams'.
```

Since `URLSearchParams` is defined on `globalThis` it doesn't have to be imported from `url`

**Comment by szmarczak** [comment]
> Since URLSearchParams is defined on globalThis it doesn't have to be imported from url.

Yeah but TS is dumb and doesn't know this. You need to `import {URL} from 'url';` for now.

**Comment by szmarczak** [comment]
Regarding no 1:

```js
new URLSearchParams({a: undefined}).toString(); //=> `a=undefined`
new URLSearchParams({}).toString(); //=> ``
```

Both objects have the type `{a: string | undefined}`.

@sindresorhus What should we do about this? `for in` and remove all undefined values?

**Comment by szmarczak** [comment]
I think no 2 is moot since we have these types:

```ts
export type OptionsOfTextResponseBody = Merge<Options, {isStream?: false; resolveBodyOnly?: false; responseType?: 'text'}>;
export type OptionsOfJSONResponseBody = Merge<Options, {isStream?: false; resolveBodyOnly?: false; responseType?: 'json'}>;
export type OptionsOfBufferResponseBody = Merge<Options, {isStream?: false; resolveBodyOnly?: false; responseType: 'buffer'}>;
export type OptionsOfUnknownResponseBody = Merge<Options, {isStream?: false; resolveBodyOnly?: false}>;
export type StrictOptions = Except<Options, 'isStream' | 'responseType' | 'resolveBodyOnly'>;
```

Also if you provide generalized `Options` object, the `got` functions returns a promise resolving either with `CancelableRequest` (promise) or with a `Request` (stream). I think you're using Got 10, please update to Got 11.3.0.

**Comment by sindresorhus** [comment]
> What should we do about this? for in and remove all undefined values?

Ideally, I would have preferred to just stay with the `URLSearchParams` behavior, but I agree not allowing undefined is inconvenient and stringifying it is definitely not what you'd want. However, we'll need to clearly document the special behavior.

**Comment by szmarczak** [comment]
1. is fixed in 4dbada94e261ad890d2478c313ec3d74761b659e

**Comment by szmarczak** [comment]
2. is totally legit. You're trying to get a Promise with your options defined as `GotOptions` that define `isStream` as `boolean`, therefore the Promise cannot be guaranteed. Use `StrictOptions` instead, or `Except<GotOptions, 'isStream'>`, or `Merge<GotOptions, {isStream?: false}>`.
